File: mcp-code-intelligence-kotlin/bin/main/com/codeintel/memory/consolidation/TierConsolidator.kt
/** 4-Tier consolidation engine — promotes/demotes entries between tiers. */
package com.codeintel.memory.consolidation

import com.codeintel.log
import com.codeintel.memory.models.MemoryTier
import com.codeintel.memory.repository.ConsolidationRepository
import com.codeintel.memory.repository.KnowledgeRepository

/** Configuration for tier promotion/demotion thresholds. */
data class ConsolidationConfig(
    val workingToEpisodicAccess: Int = 3,
    val workingToEpisodicConfidence: Double = 0.7,
    val episodicToSemanticAccess: Int = 10,
    val episodicToSemanticConfidence: Double = 0.85,
    val semanticToProceduralAccess: Int = 25,
    val semanticToProceduralConfidence: Double = 0.95,
    val demotionInactiveDays: Int = 30
)

/** Result of a consolidation run. */
data class ConsolidationResult(
    val promoted: Int,
    val demoted: Int,
    val expired: Int
)

class TierConsolidator(
    private val knowledgeRepo: KnowledgeRepository,
    private val consolidationRepo: ConsolidationRepository,
    private val config: ConsolidationConfig = ConsolidationConfig()
) {

    /** Run full consolidation cycle across all tiers. */
    fun consolidate(): ConsolidationResult {
        val fixed = fixTierMismatches()
        val promoted = promoteEligible()
        val expired = expireStale()
        log("Consolidation: fixed=$fixed, promoted=$promoted, expired=$expired")
        return ConsolidationResult(promoted + fixed, 0, expired)
    }

    /** Fix entries whose tier doesn't match their type (legacy data). */
    private fun fixTierMismatches(): Int {
        val working = knowledgeRepo.findByTier(MemoryTier.WORKING.name, limit = 1000)
        var fixed = 0
        for (entry in working) {
            val correctTier = tierForType(entry.type)
            if (correctTier != MemoryTier.WORKING.name) {
                knowledgeRepo.updateTier(entry.id, correctTier)
                consolidationRepo.logTransition(entry.id, MemoryTier.WORKING.name, correctTier, "auto:tier_fix")
                fixed++
            }
        }
        return fixed
    }

    /** Determine correct tier based on knowledge type. */
    private fun tierForType(type: String): String = when (type) {
        "REQUIREMENT", "ARCHITECTURE", "PROCEDURE", "API_DESIGN" -> MemoryTier.SEMANTIC.name
        "DECISION", "LESSON_LEARNED" -> MemoryTier.EPISODIC.name
        "ERROR_PATTERN" -> MemoryTier.EPISODIC.name
        else -> MemoryTier.WORKING.name
    }

    /** Promote entries that meet threshold criteria. */
    private fun promoteEligible(): Int {
        var count = 0
        count += promoteFromTier(
            MemoryTier.WORKING.name, MemoryTier.EPISODIC.name,
            config.workingToEpisodicAccess, config.workingToEpisodicConfidence
        )
        count += promoteFromTier(
            MemoryTier.EPISODIC.name, MemoryTier.SEMANTIC.name,
            config.episodicToSemanticAccess, config.episodicToSemanticConfidence
        )
        count += promoteFromTier(
            MemoryTier.SEMANTIC.name, MemoryTier.PROCEDURAL.name,
            config.semanticToProceduralAccess, config.semanticToProceduralConfidence
        )
        return count
    }

    private fun promoteFromTier(fromTier: String, toTier: String, minAccess: Int, minConf: Double): Int {
        val candidates = consolidationRepo.findPromotionCandidates(fromTier, minAccess, minConf)
        for (id in candidates) {
            knowledgeRepo.updateTier(id, toTier)
            consolidationRepo.logTransition(id, fromTier, toTier, "auto:threshold_met")
        }
        return candidates.size
    }

    /** Remove expired entries. */
    private fun expireStale(): Int {
        // Only expire WORKING tier entries with expires_at set
        val working = knowledgeRepo.findByTier(MemoryTier.WORKING.name, limit = 500)
        var expired = 0
        for (entry in working) {
            if (entry.expiresAt != null && isExpired(entry.expiresAt)) {
                knowledgeRepo.delete(entry.id)
                expired++
            }
        }
        return expired
    }

    private fun isExpired(expiresAt: String): Boolean {
        return try {
            val now = java.time.LocalDateTime.now().toString()
            expiresAt < now
        } catch (_: Exception) { false }
    }
}

File: mcp-code-intelligence-nodejs/src/memory/tier-consolidator.ts
/**
 * TierConsolidator — promotes/demotes entries between 4 tiers.
 * WORKING → EPISODIC → SEMANTIC → PROCEDURAL
 */

import { ConsolidationResult } from './models.js';
import { KnowledgeRepository } from './knowledge-repo.js';
import { ConsolidationRepository } from './consolidation-repo.js';

interface ConsolidationConfig {
  workingToEpisodicAccess: number;
  workingToEpisodicConfidence: number;
  episodicToSemanticAccess: number;
  episodicToSemanticConfidence: number;
  semanticToProceduralAccess: number;
  semanticToProceduralConfidence: number;
}

const DEFAULT_CONFIG: ConsolidationConfig = {
  workingToEpisodicAccess: 3,
  workingToEpisodicConfidence: 0.7,
  episodicToSemanticAccess: 10,
  episodicToSemanticConfidence: 0.85,
  semanticToProceduralAccess: 25,
  semanticToProceduralConfidence: 0.95,
};

export class TierConsolidator {
  private readonly knowledgeRepo: KnowledgeRepository;
  private readonly consolidationRepo: ConsolidationRepository;
  private readonly config: ConsolidationConfig;

  constructor(
    knowledgeRepo: KnowledgeRepository,
    consolidationRepo: ConsolidationRepository,
    config = DEFAULT_CONFIG
  ) {
    this.knowledgeRepo = knowledgeRepo;
    this.consolidationRepo = consolidationRepo;
    this.config = config;
  }

  /** Run full consolidation cycle. */
  consolidate(): ConsolidationResult {
    const fixed = this.fixTierMismatches();
    const promoted = this.promoteEligible();
    const expired = this.expireStale();
    return { promoted: promoted + fixed, demoted: 0, expired };
  }

  /** Fix entries whose tier doesn't match their type (legacy data). */
  private fixTierMismatches(): number {
    const working = this.knowledgeRepo.findByTier('WORKING', 1000);
    let fixed = 0;
    for (const entry of working) {
      const correctTier = this.tierForType(entry.type);
      if (correctTier !== 'WORKING') {
        this.knowledgeRepo.updateTier(entry.id, correctTier);
        this.consolidationRepo.logTransition(entry.id, 'WORKING', correctTier, 'auto:tier_fix');
        fixed++;
      }
    }
    return fixed;
  }

  /** Determine correct tier based on knowledge type. */
  private tierForType(type: string): string {
    switch (type) {
      case 'REQUIREMENT': case 'ARCHITECTURE': case 'PROCEDURE': case 'API_DESIGN':
        return 'SEMANTIC';
      case 'DECISION': case 'LESSON_LEARNED': case 'ERROR_PATTERN':
        return 'EPISODIC';
      default:
        return 'WORKING';
    }
  }

  private promoteEligible(): number {
    let count = 0;
    count += this.promoteFromTier('WORKING', 'EPISODIC',
      this.config.workingToEpisodicAccess, this.config.workingToEpisodicConfidence);
    count += this.promoteFromTier('EPISODIC', 'SEMANTIC',
      this.config.episodicToSemanticAccess, this.config.episodicToSemanticConfidence);
    count += this.promoteFromTier('SEMANTIC', 'PROCEDURAL',
      this.config.semanticToProceduralAccess, this.config.semanticToProceduralConfidence);
    return count;
  }

  private promoteFromTier(from: string, to: string, minAccess: number, minConf: number): number {
    const candidates = this.consolidationRepo.findPromotionCandidates(from, minAccess, minConf);
    for (const id of candidates) {
      this.knowledgeRepo.updateTier(id, to);
      this.consolidationRepo.logTransition(id, from, to, 'auto:threshold_met');
    }
    return candidates.length;
  }

  private expireStale(): number {
    const working = this.knowledgeRepo.findByTier('WORKING', 500);
    let expired = 0;
    const now = new Date().toISOString();
    for (const entry of working) {
      if (entry.expires_at && entry.expires_at < now) {
        this.knowledgeRepo.delete(entry.id);
        expired++;
      }
    }
    return expired;
  }
}
